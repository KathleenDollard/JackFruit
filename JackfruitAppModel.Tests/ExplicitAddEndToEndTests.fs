namespace Generator

open Microsoft.CodeAnalysis;
open Xunit
open ApprovalTests.Reporters
open System
open Generator.Tests.UtilsForTests
open ApprovalTests
open System.Linq
open Generator.SourceGenerator
open Generator
open Generator.Tests.MapExplicitAddData
open Generator.ExplicitAdd

type ExplicitAddTestGenerator() =
    inherit Jackfruit.ExplicitAddGenerator()

    override _.CodeModelBuilder commandDef =
        BuildNewerCliCodeModel.OutputCommandWrapper commandDef

//[<Generator>]
type ExplicitAddEndToEndTests() =
    let RunCliModelGenerator (inputCompilation: Compilation) generator =
        RunGenerator generator inputCompilation

    let RunTest (data: Data) generator =
        let compilation = CreateCompilation data.CliCode
        let inputDiagnostics = compilation.GetDiagnostics()
        if not (inputDiagnostics.IsEmpty) then invalidOp "Input code is invalid"
        let (driver, outputCompilation, diagnostics) = RunCliModelGenerator compilation generator
        if not (inputDiagnostics.IsEmpty) then invalidOp "Generation failed"
        let newTree = outputCompilation.SyntaxTrees.Skip(1).SingleOrDefault()
        Assert.NotNull newTree
        Approvals.Verify(newTree.ToString())   

     
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``No commands``() =
        RunTest NoMapping (Jackfruit.ExplicitAddGenerator())
        
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``One simple command``() =
        RunTest OneMapping (Jackfruit.ExplicitAddGenerator())
    
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``Three commands``() =
        RunTest ThreeMappings (Jackfruit.ExplicitAddGenerator())

    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``No commands with test generator``() =
        RunTest NoMapping (ExplicitAddTestGenerator())
   
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``One simple command with test generator``() =
        RunTest OneMapping (ExplicitAddTestGenerator())
    
    [<Fact>]
    [<UseReporter(typeof<DiffReporter>)>]
    member _.``Three commands with test generator``() =
        RunTest ThreeMappings (ExplicitAddTestGenerator())
