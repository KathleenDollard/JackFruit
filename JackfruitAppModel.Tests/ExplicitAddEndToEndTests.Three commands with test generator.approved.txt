using System;
using System.CommandLine;
using System.CommandLine.Invocation;
using System.Threading.Tasks;
namespace CliDefinition
{
   
   public class OriginalSeriesCli
   {
      public Command Command {get; set;}
      public OriginalSeriesCli()
      {
         Command = new Command("OriginalSeries");
         kirkOption = new Option<string>("kirk");
         Command.Add(kirkOption);
         spockOption = new Option<string>("spock");
         Command.Add(spockOption);
         uhuraOption = new Option<string>("uhura");
         Command.Add(uhuraOption);
         Command.SetHandler(CliApp.Handlers.OriginalSeries, kirkOption, spockOption, uhuraOption);
      }
      public Option<string> kirkOption {get; set;}
      public Option<string> spockOption {get; set;}
      public Option<string> uhuraOption {get; set;}
   }
   
   public class NextGenerationCli
   {
      public Command Command {get; set;}
      public NextGenerationCli()
      {
         Command = new Command("NextGeneration");
         picardOption = new Option<string>("picard");
         Command.Add(picardOption);
         Command.SetHandler(CliApp.Handlers.NextGeneration, picardOption);
      }
      public Option<string> picardOption {get; set;}
   }
   
   public class VoyagerCli
   {
      public Command Command {get; set;}
      public VoyagerCli()
      {
         Command = new Command("Voyager");
         janewayOption = new Option<string>("janeway");
         Command.Add(janewayOption);
         chakotayOption = new Option<string>("chakotay");
         Command.Add(chakotayOption);
         torresOption = new Option<string>("torres");
         Command.Add(torresOption);
         tuvokOption = new Option<bool>("tuvok");
         Command.Add(tuvokOption);
         sevenOfNineOption = new Option<string>("sevenOfNine");
         Command.Add(sevenOfNineOption);
         Command.SetHandler(CliApp.Handlers.Voyager, janewayOption, chakotayOption, torresOption, tuvokOption, sevenOfNineOption);
      }
      public Option<string> janewayOption {get; set;}
      public Option<string> chakotayOption {get; set;}
      public Option<string> torresOption {get; set;}
      public Option<bool> tuvokOption {get; set;}
      public Option<string> sevenOfNineOption {get; set;}
   }
}
